masonry_core/src/app/render_root.rs

```diff
@@ -287,6 +306,80 @@ pub(crate) struct InspectorState {
 }
 
 impl RenderRoot {
+    /// Compute a focus fallback according to the provided `policy`.
+    ///
+    /// Returns `Some(id)` if a suitable widget is found, otherwise `None`.
+    pub fn compute_focus_fallback(&self, policy: FocusFallbackPolicy) -> Option<WidgetId> {
+        let base = self.layer_root_id(0);
+        match &policy {
+            FocusFallbackPolicy::AnchorSubtreeFirst(inner) => {
+                if let Some(anchor) = self.global_state.focus_anchor
+                    && self.is_still_interactive(anchor)
+                    && let Some(id) = self.match_policy_at(inner, anchor)
+                {
+                    return Some(id);
+                }
+                self.match_policy_at(inner, base)
+            }
+            FocusFallbackPolicy::FirstTextInput
+            | FocusFallbackPolicy::FirstFocusable
+            | FocusFallbackPolicy::Tagged(_) => self.match_policy_at(&policy, base),
+        }
+    }
+
+    /// Depth-first pre-order search from `start` for first widget whose `predicate` returns true.
+    fn find_first_matching(
+        &self,
+        start: WidgetId,
+        predicate: impl Fn(&WidgetRef<'_, dyn Widget>) -> bool,
+    ) -> Option<WidgetId> {
+        let mut stack = vec![self.get_widget(start)?];
+        while let Some(node) = stack.pop() {
+            if predicate(&node) {
+                return Some(node.id());
+            }
+            let mut children = node.children();
+            for child in children.drain(..).rev() {
+                stack.push(child);
+            }
+        }
+        None
+    }
+
+    /// Match a policy starting at `start`.
+    fn match_policy_at(&self, policy: &FocusFallbackPolicy, start: WidgetId) -> Option<WidgetId> {
+        match policy {
+            FocusFallbackPolicy::FirstTextInput => self.find_first_matching(start, |node| {
+                let ctx = node.ctx();
+                ctx.accepts_text_input() && !ctx.is_disabled() && !ctx.is_stashed()
+            }),
+            FocusFallbackPolicy::FirstFocusable => self.find_first_matching(start, |node| {
+                let ctx = node.ctx();
+                ctx.accepts_focus() && !ctx.is_disabled() && !ctx.is_stashed()
+            }),
+            FocusFallbackPolicy::Tagged(name) => {
+                let id = self.global_state.widget_tags.get(name).copied()?;
+                if self.is_still_interactive(id) {
+                    Some(id)
+                } else {
+                    None
+                }
+            }
+            FocusFallbackPolicy::AnchorSubtreeFirst(_) => None,
+        }
+    }
+
+    /// Compute and set a focus fallback according to the provided `policy`.
+    ///
+    /// Returns `true` if a fallback was found and set.
+    pub fn set_focus_fallback_policy(&mut self, policy: FocusFallbackPolicy) -> bool {
```

Comment:
I think this function name probably needs to change, as again, "policy" to me strongly indicates that this is continuous. I'm not sure what the oneshot name for this would be, though.